clis/tcli/src/bin/dtcli/main.rs

```diff
@@ -0,0 +1,187 @@
+    input.insert(0, "dtcli".to_owned());
+
+    println!("{}", TelioCliInterface::get_help(input)?);
+
+    Ok(())
+}
+
+/// Main function of the API. This is where all the API's decision making happens.
+///
+/// # Arguments
+///
+/// * `args` - Command line arguments for the API (already parsed).
+fn run_api(args: Args) -> Result<()> {
+    // We don't want to start the daemon if all we want is to print the help message.
+    // But we also have to retrieve the help message from the TCLI parser.
+    if args.help {
+        return print_help(args.input);
+    }
+
+    match args.input.first().map(|s| s.as_str()) {
+        // If no arguments are supplied default behavior is starting the daemon.
+        None => {
+            if Daemon::is_running()? {
+                eprintln!("DTCLI daemon is already running, stop it by calling `dtcli quit`");
+            } else {
+                start_daemon(args.features.clone())?;
+            }
+        }
+        Some("quit") => {
+            if Daemon::is_running()? {
+                let response = DaemonIPC::send_command(&get_ipc_socket_path()?, "quit")?;
+
+                if response.as_str() == "OK" {
+                    println!("DTCLI daemon stopped");
+                } else {
+                    eprintln!("Error while trying to stop DTCLI daemon, please check if DTCLI daemon is still running and close it manually");
+                }
+            } else {
+                eprintln!("DTCLI daemon is not running");
+            }
+        }
+        Some(_) => {
+            if !Daemon::is_running()? {
+                start_daemon(args.features.clone())?;
+            } else if args.features.is_some() {
+                eprintln!("Features cannot be set while DTCLI daemon is running.");
+            }
+            let response = DaemonIPC::send_command(&get_ipc_socket_path()?, &args.input.join(" "))?;
+            println!("{}", response);
+        }
+    };
+
+    Ok(())
+}
+
+/// Starts the DTCLI daemon which will run actual Telio. For Unix systems uses forking. Decided not to
+/// separate api and dtcli.d into separate binaries in the project, because that added a lot of extra work.
+/// Like having to build the dtcli.d binary manually before running the api and didn't really pay off the extra effort.
+///
+/// # Arguments
+///
+/// * `features` - Optional string containing the features specification for Telio in JSON.
+fn start_daemon(features: Option<String>) -> Result<()> {
+    println!("Starting DTCLI daemon...");
+    match Daemon::start() {
+        Ok(ProcessType::Api) => {
+            // Verifying that the server had been started, but we have no way of synchronizing
+            // the verification with the daemon, so using retries instead.
+            DaemonIPC::send_command_with_retries(
+                &get_ipc_socket_path()?,
+                "status",
+                &[
+                    Duration::from_millis(500),
```

Comment:
How were these delay values chosen ?

Reply:
Experimentally by testing how long it takes for the IPC listener and Telio. There wasn't really any numbers on which I could rely on nor was this a big risk of failure point.

Reply:
I'll add a comment about it

Reply:
Mm. Then these numbers could be incrementally increasing.

Reply:
I'd rather keep them as they are, since the daemon never starts sooner than 500 ms and the largest possibility of it starting is between 500 and 1100 ms. I left the longer delay at the end in case there's some edge case where it takes longer for some reason.